text/0000-install-pin-npm.md

```diff
@@ -0,0 +1,65 @@
+- Feature Name: install_pin_npm
+- Start Date: 2019-10-14
+- RFC PR: (leave this empty)
+- Volta Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Support installation and pinning of arbitrary `npm` versions, decoupled from the selected Node version.
+
+# Motivation
+[motivation]: #motivation
+
+For package managers, we currently only support using an arbitrary version of `yarn`, but not `npm`. For `npm`, users are forced to use the version that is automatically bundled with Node, meaning they can't upgrade or downgrade without changing the Node version. We should support `npm` to the same level that we support `yarn` and give users full control over the version of their package manager.
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+There shouldn't be anything new we need to document / teach users, as the behavior already exists for Node and Yarn. The only change is that we will support the `volta install npm` and `volta pin npm` commands in the same way that we support `volta install yarn` and `volta pin yarn`.
+
+# Details
+[details]: #details
+
+## Installation
+
+Unlike Yarn, `npm` doesn't provide a pre-packaged tarball for download. Instead, the rely on using the existing `npm` installation that is provided with Node to do any upgrades. This is very similar to how we handle global 3rd-party packages, so we should be able to leverage a similar approach for installing `npm`. That does mean, however, that in the same way we require Node to be available to install packages, we will also require Node to be available to install a custom version of `npm`.
+
+It appears we can also skip the `npm install` step, as the `npm` package tarball comes with all of the dependencies already included.
+
+## Running
+
+Currently, to execute `npm`, we add the Node bin directory to the PATH and then call `npm`, relying on the fact that `npm` is bundled with Node, so `npm` is available within the Node bin directory. To support arbitrary versions of `npm`, we will need to take the approach that is used for packages and call the `npm-cli.js` file directly. Since this isn't an arbitrary package, but will always be `npm`, we can likely shortcut the configuration files and detection of binaries that we need for packages and use the known path to the file.
```

Comment:
Does this mean that we will defer to the "project local" version if the project depends on `npm`? That seems at the same time "natural" and "weird" 😝

Reply:
Probably not, though we could look at doing so. My thought was that the logic would be similar, but still different enough to be distinct, so they would have different behavior. Specifically, they would be a sort of melding of the logic for `yarn` and the logic for packages.